src/main/java/org/testng/xml/Parser.java

```diff
@@ -266,7 +270,13 @@ public static boolean hasFileScheme(String uri) {
     return result;
   }
 
-
+  private static URI constructURI(String text) {
+    try {
```

Comment:
```java
try {
  return new URI(text);
} catch (URISyntaxException e) {
  return null;
}
```

Reply:
@juherr - How is this different from using the factory method  `URI.create()` ? Can you please elaborate ? The factory method also calls into this constructor only.  Since in both cases we are gobbling exceptions and just returning `null`, I wasn't able to figure out what advantage the constructor invocation would have over the factory method. Please help me understand.

Reply:
I think that a double catch is a good pattern and according to the doc:

> This method is provided for use in situations where it is known that the given string is a legal URI

https://docs.oracle.com/javase/8/docs/api/java/net/URI.html#create-java.lang.String-

Reply:
Yes I read that. But in this case, we aren't doing anything with the outcome. We are merely gobbling the exception and then returning a `null`. So IMO it shouldn't matter. Moreover the factory method is calling this constructor only. Here's the code for the `create` method.

```java
public static URI create(String str) {
    try {
        return new URI(str);
    } catch (URISyntaxException x) {
        throw new IllegalArgumentException(x.getMessage(), x);
    }
}
```

Reply:
`constructURI` is just catching `IllegalArgumentException` and returns null.
So, why not catching `URISyntaxException` and returns null?

Reply:
`constructURI` catches `Exception` and not `IllegalArgumentException`. Please see [here](https://github.com/krmahadevan/testng/blob/f8c45ccd68050532f03182f58d664a46a6af6a7e/src/main/java/org/testng/xml/Parser.java#L274-L279)

Reply:
But exception IllegalArgumentException, what are the possible exceptions?

Reply:
Skimming through the code of `URI` class, I can see that it can throw either `NullPointerException` (or) `IllegalArgumentException` if we use `URI.create()`. If we use `new URI(str)`, it can throw either `NullPointerException` (or) `URISyntaxException`.

Reply:
In our case, str cannot be null (you can enforce it with a `@NotNull` if you want)

Reply:
@juherr - Agreed. But I still don't understand the point you are driving. Why shouldn't we be using `URI.create()` and how is it not better when compared to using `new URI(str)` especially when our use case is pretty simple viz., *Attempt at creating a URI. If the attempt fails, then fall back to using a null value*

I am still not convinced why a change is warranted here. So please help me understand that part.

Reply:
First, the doc is explicit enough:

> or use in situations where it is known that the given string is a legal URI

Which is not our case.

Then, we know that exceptions cost and the current code is like:

```java
try {
    try {
        return new URI(str);
    } catch (URISyntaxException x) {
        throw new IllegalArgumentException(x.getMessage(), x);
    }
} catch (IllegalArgumentException e) {
    return null;
}
```

Weird, isn't it?

Reply:
I dont see how its weird especially when an exception is going to be created only when either the URI is not valid and we end up having to gobble it up and give back a `null`.Yes I know the cost of creating an exception (especially when it has to build the stacktrace internally), but I am not completely convinced that `URI.create()` is a bad option when compared to using `new URI(str)` especially when the former is going to be calling in the latter only. So I guess I will just agree to disagree on this one.

Reply:
The question is why do you do 2 catches when only 1 is possible?